aten/src/ATen/xpu/XPUContext.cpp

```diff
@@ -0,0 +1,91 @@
+#include <ATen/xpu/XPUContext.h>
+#include <c10/util/CallOnce.h>
+#include <c10/util/Exception.h>
+
+#include <cmath>
+#include <deque>
+#include <mutex>
+#include <vector>
+
+namespace at::xpu {
+namespace {
+
+/*
+ * Currently, there is one device properties pool containing the information and
+ * capability about each compute-device.
+ *
+ * Device properties are lazily initialized when the first time properties are
+ * requested for a device.
+ */
+DeviceIndex num_gpus = -1;
+c10::once_flag init_flag;
+std::deque<c10::once_flag> device_prop_flags;
+std::vector<DeviceProp> device_properties;
+
+std::deque<c10::once_flag> device_global_idx_flags;
+std::vector<int> device_global_idxs;
+
+void initXPUContextVectors() {
+  num_gpus = c10::xpu::device_count();
+  device_prop_flags.resize(num_gpus);
+  device_properties.resize(num_gpus);
+  device_global_idx_flags.resize(num_gpus);
+  device_global_idxs.resize(num_gpus);
+}
+
+void initDeviceProperty(DeviceIndex device) {
+  DeviceProp device_prop;
+  c10::xpu::get_device_properties(&device_prop, device);
+  device_properties[device] = device_prop;
+}
+
+void initDeviceGlobalIdx(DeviceIndex device) {
+  sycl::device& raw_device = c10::xpu::get_raw_device(device);
+  // Get all SYCL devices associated with the SYCL platform.
+  auto devices = sycl::device::get_devices();
+  auto match_device = [raw_device](const auto& dev) -> bool {
+    return raw_device == dev;
+  };
+  auto it = std::find_if(devices.begin(), devices.end(), match_device);
+  TORCH_CHECK(
+      it != devices.end(), "Cant't find the global index of XPU device.");
+  device_global_idxs[device] =
+      static_cast<int>(std::distance(devices.begin(), it));
+}
+
+inline void check_device(DeviceIndex device) {
+  TORCH_CHECK(
+      device >= 0 && device < num_gpus,
+      "device is out of range, device is ",
+      static_cast<int>(device),
+      ", total number of device is ",
+      static_cast<int>(num_gpus),
```

Comment:
Why do you need to cast here?
```suggestion
      num_gpus,
```

Reply:
`device` and `num_gpus` are `DeviceIndex` types, which is `int8_t`. `int8_t` is similar as `signed char` and `std::cout` can not correctly print its value. refer to https://stackoverflow.com/questions/7587782/why-is-stdcout-not-printing-the-correct-value-for-my-int8-t-number
According to the existing PyTorch [code](https://github.com/pytorch/pytorch/blob/1872834247084d76544c95e04fe2ec3370bac87d/c10/core/DeviceType.cpp#L58) in `c10/core/DeviceType.cpp`, I will change `static_cast<int>(num_gpus)` to `static_cast<int16_t>(num_gpus)`.